File: task_3/3.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

int min_edges = INT_MAX;
int best_mask = 0;

int count_cross_edges(const vector<vector<int>>& A, int mask, int n) {
    int count = 0;
    for (int i = 0; i < n; ++i)
        for (int j = i + 1; j < n; ++j)
            if (A[i][j]) {
                bool in_diff_groups = ((mask >> i) & 1) != ((mask >> j) & 1);
                if (in_diff_groups) count++;
            }
            return count;
}

// Рекурсивная функция перебора масок
void search(int bit, int mask, const vector<vector<int>>& A, int n) {
    if (bit == n) {
        // Пропускаем полные группы (все 0 или все 1)
        if (mask == 0 || mask == (1 << n) - 1) return;

        int edges = count_cross_edges(A, mask, n);
        if (edges < min_edges) {
            min_edges = edges;
            best_mask = mask;
        }
        return;
    }

    // Рекурсивно пробуем: страна `bit` в группе 0 и в группе 1
    search(bit + 1, mask, A, n); // Страна в группе 0
    search(bit + 1, mask | (1 << bit), A, n); // Страна в группе 1
}

int main() {
    int n;
    cout << "Введите количество стран: ";
    cin >> n;

    vector<vector<int>> A(n, vector<int>(n));
    cout << "Введите матрицу смежности (" << n << "x" << n << "):\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> A[i][j];

    search(0, 0, A, n); // Запуск перебора с пустой маской

    cout << "Минимальное количество связей между группами: " << min_edges << endl;

    cout << "Группа 1: ";
    for (int i = 0; i < n; ++i)
        if ((best_mask >> i) & 1)
            cout << i + 1 << " ";
    cout << endl;

    cout << "Группа 2: ";
    for (int i = 0; i < n; ++i)
        if (!((best_mask >> i) & 1))
            cout << i + 1 << " ";
    cout << endl;

    return 0;
}

File: task_3/3.go
package main

import (
	"fmt"
	"math"
)

var (
	minEdges  = math.MaxInt
	bestMask  int
	adjMatrix [][]int
)

func countCrossEdges(mask, n int) int {
	count := 0
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if adjMatrix[i][j] == 1 {
				inDiffGroups := ((mask>>i)&1) != ((mask>>j)&1)
				if inDiffGroups {
					count++
				}
			}
		}
	}
	return count
}

func search(bit, mask, n int) {
	if bit == n {
		if mask == 0 || mask == (1<<n)-1 {
			return // Пропускаем пустую или полную группу
		}
		edges := countCrossEdges(mask, n)
		if edges < minEdges {
			minEdges = edges
			bestMask = mask
		}
		return
	}

	// Рекурсивно пробуем: страна `bit` в группе 0 и в группе 1
	search(bit+1, mask, n) // В группе 0
	search(bit+1, mask|(1<<bit), n) // В группе 1
}

func main() {
	var n int
	fmt.Print("Введите количество стран: ")
	fmt.Scan(&n)

	adjMatrix = make([][]int, n)
	fmt.Printf("Введите матрицу смежности (%dx%d):\n", n, n)
	for i := range adjMatrix {
		adjMatrix[i] = make([]int, n)
		for j := range adjMatrix[i] {
			fmt.Scan(&adjMatrix[i][j])
		}
	}

	search(0, 0, n)

	fmt.Println("Минимальное количество связей между группами:", minEdges)

	fmt.Print("Группа 1: ")
	for i := 0; i < n; i++ {
		if (bestMask>>i)&1 == 1 {
			fmt.Print(i+1, " ")
		}
	}
	fmt.Println()

	fmt.Print("Группа 2: ")
	for i := 0; i < n; i++ {
		if (bestMask>>i)&1 == 0 {
			fmt.Print(i+1, " ")
		}
	}
	fmt.Println()
}
